ir/function.cpp

```diff
@@ -428,5 +438,115 @@ void DomTree::printDot(std::ostream &os) const {
   os << "}\n";
 }
 
+
+void LoopAnalysis::getDepthFirstSpanningTree() {
+  std::map<const BasicBlock *, std::vector<const BasicBlock *>> edges;
+  unsigned current = 0;
+  for (const auto &[src, dst, instr] : cfg) {
+    edges[&src].push_back(&dst);
+  }
+
+  std::function<void(const BasicBlock *)> dfs = [&](const BasicBlock *bb) {
+    node.push_back(bb);
+    number[bb] = current++;
+    for (auto tgt : edges [bb]) {
+      if (number.find(tgt) == number.end()) {
+        dfs(tgt);
+      }
+    }
+    last[number[bb]] = current - 1;
+  };
+
+  dfs(&f.getFirstBB());
+}
+
+void LoopAnalysis::printNodeType(ostream &os, const NodeType &t) const {
+  switch (t) {
+  case nonheader: os<<"n."; break;
+  case self:      os<<"s."; break;
+  case reducible: os<<"r."; break;
+  }
+}
+
+// Tarjan, R. (1974).
+// Testing Flow Graph Reducibility.
```

Comment:
Reason for citing this rather than Havlak's: Both paper are describing the same algorithm except that Havlak's has extra steps for handling irreducible loops. As we won't support irreducible loops, it would be better to cite the original paper, imo.

We could change it back to Havlak's if someday we implemented the full algorithm.

Reply:
you should cite both papers and say the same thign in the comment that you said here!